### PR TITLE
[WIP] Auto Mixed Precision for dygraph, test=develop

### PR DESCRIPTION
**Pull request by zhiqiu**
two phases.

- cutocast tensor data type
```
with fluid.dygraph.amp.autocast():
    loss = model(inputs)
```
For each executed operator in the model, the AutoCast module will automatically decide which data type is better to use, i.e., float16 or float32. The decision is made with white_op_list (which contains the operators that can adopt float16 calculation to accelerate and are considered numerically-safe) and black_op_list (which contains the operators that are considered numerically-safe using float16).

- scale loss

### CODE REVIEW

File: paddle/fluid/imperative/auto_cast.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  inline ~AutoCastGuard() { tracer_->SetEnableAutoCast(pre_mode_); }
+
+  AutoCastGuard(const AutoCastGuard& guard) = delete;
+  AutoCastGuard& operator=(const AutoCastGuard& guard) = delete;
+
+ private:
+  std::shared_ptr<Tracer> tracer_;
+  bool guard_mode_;
```
Comment (wzzju) [inline paddle/fluid/imperative/auto_cast.h:45]: Is `guard_mode_ ` an useless var?

File: paddle/fluid/imperative/auto_cast.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#include "paddle/fluid/imperative/tracer.h"
+#include "paddle/fluid/imperative/type_defs.h"
+
+namespace paddle {
+namespace imperative {
+
+// NOTE(zhiqiu): AutoCastGuard is used for RAII.
+class AutoCastGuard {
+ public:
+  explicit inline AutoCastGuard(std::shared_ptr<Tracer> tracer, bool guard_mode)
```
Comment (wzzju) [inline paddle/fluid/imperative/auto_cast.h:30]: The `AutoCastGuard ` is implemented in the class definition, so it is an inline method. The key work `inline` is redundancy. `explicit` is useless here, because `guard_mode ` has no default value.

File: paddle/fluid/imperative/auto_cast.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      : tracer_(tracer), guard_mode_(guard_mode) {
+    pre_mode_ = tracer_->IsAutoCastEnabled();
+    if (pre_mode_ != guard_mode) {
+      tracer_->SetEnableAutoCast(guard_mode_);
+    }
+  }
+
+  inline ~AutoCastGuard() { tracer_->SetEnableAutoCast(pre_mode_); }
```
Comment (wzzju) [inline paddle/fluid/imperative/auto_cast.h:38]: `~AutoCastGuard()` is the same with above, the `inline` is redundancy.

File: paddle/fluid/imperative/auto_cast.cc [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    for (const auto& op : pair.second) {
+      if (op->DataType() == framework::proto::VarType::FP32) {
+        dst_type = op->DataType();
+      }
+    }
+  }
+  return dst_type;
+}
+
+NameVarBaseMap AutoCastInputs(const std::string& op_type,
```
Comment (wzzju) [inline paddle/fluid/imperative/auto_cast.cc:111]: Is the `AutoCastInputs` function just called once? If not so, I think you can reduce the number of function call `CastTo*` by checking the condition in advance.

File: paddle/fluid/imperative/layer.cc [hunk 5]
```diff
@@ -336,7 +336,7 @@ static void ClearNoNeedBufferInputs(OpBase* op) {
   for (auto& slot : no_need_buffer_slots) {
     auto iter = ins->find(slot);
     if (iter == ins->end()) continue;
-    VLOG(2) << "Clear data buffer of " << slot << " in " << op->Type();
+    VLOG(3) << "Clear data buffer of " << slot << " in " << op->Type();
```
Comment (wzzju) [inline paddle/fluid/imperative/layer.cc:339]: What's the motivation of the change here?

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 6]
```diff
@@ -264,13 +267,40 @@ std::shared_ptr<VariableWrapper> SelectedRowsMerge(
       framework::DataTypeToString(data_type)));
 }
 
+// NOTE(zhiqiu): call elementwise_add to support fp16 add cuda device.
+void TensorAddFP16(std::shared_ptr<VariableWrapper> src, VariableWrapper* dst) {
+  auto op =
+      framework::OpRegistry::CreateOp("elementwise_add", {}, {}, {}, false);
+  // NOTE(zhiqiu): do not delete dst when y is out of scope.
+  auto y = std::shared_ptr<VariableWrapper>(dst, [](VariableWrapper*) {});
```
Comment (wzzju) [inline paddle/fluid/imperative/gradient_accumulator.cc:275]: It is an interesting piece of code. 👍

File: paddle/fluid/operators/amp/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+include(operators)
+register_operators()
```
Comment (wzzju) [inline paddle/fluid/operators/amp/CMakeLists.txt:2]: What is it doing here ? 🤔

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        "(Tensor) The input tensors of amp_check_finite_and_scale operator.")
+        .AsDuplicable();
+    AddInput("Scale",
+             "(Tensor) 1-dim tensor, the scale of amp_check_finite_and_scale "
+             "operator.");
+    AddOutput("Out",
+              "(Tensor) The scaled output tensor of "
+              "amp_check_finite_and_unscale operator.")
+        .AsDuplicable();
+    AddOutput("FoundInfinite",
```
Comment (wzzju) [inline paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cc:65]: Why not use `bool` instead of `int`?

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    for (size_t i = 0; i < xs.size(); ++i) {
+      const auto* x = xs[i];
+      auto* out = outs[i];
+      const T* x_data = x->data<T>();
+      T* out_data = out->mutable_data<T>(dev_ctx.GetPlace());
+
+      int num = x->numel();
+      int block = 512;
+      int grid = (num + block - 1) / block;
+      VLOG(3) << "lanuch kernel";
```
Comment (wzzju) [inline paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu:59]: `lanuch` --> `launch`

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      AmpCheckFiniteAndScale<T><<<grid, block, 0, dev_ctx.stream()>>>(
```
Comment (wzzju) [inline paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu:60]: Launching only one CUDA kernel maybe enough.

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.h [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+#include "paddle/fluid/operators/isfinite_op.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename DeviceContext, typename T>
+class AmpCheckFiniteAndScaleKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    // auto& dev_ctx = ctx.template device_context<DeviceContext>();
```
Comment (wzzju) [inline paddle/fluid/operators/amp/amp_check_finite_and_scale_op.h:28]: Have it not finished yet ?

File: paddle/fluid/pybind/op_function_generator.cc [hunk 12]
```diff
@@ -76,6 +79,9 @@ std::map<std::string, std::set<std::string>> op_passing_outs_map = {
     {"matmul", {"Out"}},
     {"fake_quantize_dequantize_moving_average_abs_max",
      {"OutScale", "OutAccum", "OutState"}},
+    {"amp_check_finite_and_scale", {"Out", "FoundInfinite"}},
+    {"fake_quantize_dequantize_moving_average_abs_max",
+     {"OutScale", "OutAccum", "OutState"}},
```
Comment (wzzju) [inline paddle/fluid/pybind/op_function_generator.cc:84]: Please don't modify the `fake_quantize_dequantize_moving_average_abs_max ` op here.

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+from ...wrapped_decorator import signature_safe_contextmanager, wrap_decorator
+from paddle.fluid import core
+import contextlib
+from ...framework import Variable, in_dygraph_mode, OpProtoHolder, Parameter, _dygraph_tracer, dygraph_only
```
Comment (wzzju) [inline python/paddle/fluid/dygraph/amp/auto_cast.py:19]: Use absolute import as the unified import way maybe a good choice.

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        _white_list, _black_list = _update_list(custom_white_list,
+                                                custom_black_list)
+
+    if tracer:
+        original_val = tracer._enable_autocast
+        original_white_list, original_black_list = tracer._get_amp_op_list()
+
+        tracer._enable_autocast = enable
+        tracer._set_amp_op_list(_white_list, _black_list)
+        original_white_list, original_black_list = tracer._get_amp_op_list()
```
Comment (wzzju) [inline python/paddle/fluid/dygraph/amp/auto_cast.py:102]: Is this an error here? You can not recover the original context after this.

File: python/paddle/fluid/tests/unittests/test_imperative_check_nan_inf.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import os
+os.environ['FLAGS_imperative_check_nan_inf'] = '1'
```
Comment (wzzju) [inline python/paddle/fluid/tests/unittests/test_imperative_check_nan_inf.py:18]: Where is the `FLAGS_imperative_check_nan_inf` definition? I just see the usage here.

File: python/paddle/fluid/dygraph/amp/loss_scaler.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        if isinstance(output, core.VarBase):
+            return output * self._scale
+        elif isinstance(outputs, list) or isinstance(outputs, tuple):
+            return [self.scale(var) for var in output]
+        else:
+            raise ValueError(
+                "output must be a Variable or an list/tuple of Variables")
+
+    @dygraph.no_grad
+    def unscale_(self, optimizer):
```
Comment (wzzju) [inline python/paddle/fluid/dygraph/amp/loss_scaler.py:82]: If `unscale_` is not a public API, please use `_unscale` instead.

File: paddle/fluid/operators/amp/CMakeLists.txt (hunk 7 above)
Comment (zhiqiu) [reply paddle/fluid/operators/amp/CMakeLists.txt:2]: To register the new op automatically.

File: python/paddle/fluid/tests/unittests/test_imperative_check_nan_inf.py (hunk 15 above)
Comment (zhiqiu) [reply python/paddle/fluid/tests/unittests/test_imperative_check_nan_inf.py:18]: Defined in `imperative/flags.cc`.

File: python/paddle/fluid/dygraph/amp/loss_scaler.py (hunk 16 above)
Comment (zhiqiu) [reply python/paddle/fluid/dygraph/amp/loss_scaler.py:82]: Actually, the underscore in suffix, i.e, `unscale_`, is used to specify that this function is `inplace`.
I will change it to `_unscale_inplace`.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 14 above)
Comment (zhiqiu) [reply python/paddle/fluid/dygraph/amp/auto_cast.py:102]: Yes, this line is for debug, I will remove it.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 13 above)
Comment (zhiqiu) [reply python/paddle/fluid/dygraph/amp/auto_cast.py:19]: OK, thx.

File: paddle/fluid/pybind/op_function_generator.cc (hunk 12 above)
Comment (zhiqiu) [reply paddle/fluid/pybind/op_function_generator.cc:84]: Done

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.h (hunk 11 above)
Comment (zhiqiu) [reply paddle/fluid/operators/amp/amp_check_finite_and_scale_op.h:28]: Yes, I will add cpu kernel later.

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu (hunk 9 above)
Comment (zhiqiu) [reply paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cu:59]: Done, thx.

File: paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cc (hunk 8 above)
Comment (zhiqiu) [reply paddle/fluid/operators/amp/amp_check_finite_and_scale_op.cc:65]: Done, thx.

File: paddle/fluid/imperative/auto_cast.h (hunk 3 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/auto_cast.h:38]: Done, thx.

File: paddle/fluid/imperative/auto_cast.h (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/auto_cast.h:30]: Done, thx.

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 6 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/gradient_accumulator.cc:275]: This is rough implementation, small trick. Here is the implementation: https://github.com/PaddlePaddle/Paddle/pull/24823.

File: paddle/fluid/imperative/layer.cc (hunk 5 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/layer.cc:339]: To reduce logs, since the information of `Clear data buffer of` is not very important and can use a higher log level.

File: paddle/fluid/imperative/auto_cast.h (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/auto_cast.h:45]: Yes, It can be removed.
